### PR TITLE
DB Index changes

### PR DESCRIPTION
**Pull request by neurosis69**
I'd like to propose index changes.

1. Reduce size of is_peak index to 1 page by replacing it by partial index
    Described in https://github.com/Chia-Network/chia-blockchain/pull/9112#issuecomment-962682781

2. Remove group by/having clauses from this bluebox support query to improve performance
    Reduced execution times on the following systems:
    @i7-11700K, db at local SATA SSD: 
        old: 3.268s 
        new: 0.525s
    @i5-8250U, db at local mSATA SSD: 
        old: 110.811s
        new: 2.151s
    @RPi 4 8GB, db at USB3-connected NVMe SSD
        old: 222.072s
        new: 0.905s
    @kvm vm with 4 rzyen 5950x threads, qcow2 storage backed by 2x samsung  pm983 3.84 tb nvme's
        old: 35.6s
        new: 0.604s

### CODE REVIEW

**Comment by emlowe** [comment]
Thanks for the PR - I dunno what the WHERE BY HAVING was trying to do but seems your query is functional equavalent. There still seems to be some tabbing issue though
```
E     File "/home/runner/work/chia-blockchain/chia-blockchain/chia/wallet/wallet_block_store.py", line 69
E       await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak) where is_peak = 1")
E                                                                                                          ^
E   TabError: inconsistent use of tabs and spaces in indentation
```

**Comment by emlowe** [comment]
It wants you to run black on the files to get the formatting correct.
```
. ./activate
pip install ".[dev]"
black chia tests && mypy chia tests && flake8 chia tests
```
you can also just run black on the particular source file black.

**Comment by mariano54** [comment]
Thank you for the PR on this! Can you explain what the benchmarks are that you ran, that resulted in those large lookup times?

**Comment by neurosis69** [comment]
> 
> 
> It wants you to run black on the files to get the formatting correct.
> 
> ```
> . ./activate
> pip install ".[dev]"
> black chia tests && mypy chia tests && flake8 chia tests
> ```
> 
> you can also just run black on the particular source file

Found and fixed the trailing whitespaces, but got a lots of other warnings and pre-commit didn't allow me to commit.
So I just changed it in web repo - I hope it passes all these checks now.

**Comment by neurosis69** [comment]
> 
> 
> Thank you for the PR on this! Can you explain what the benchmarks are that you ran, that resulted in those large lookup times?

To reduce the actual output of all the heights, I put the statement as subselect and added a count.
(but I also ran it without count+subselect a couple of times, to be sure)

I used the following stmts (I think in sql trace I saw the limit of 30 to be used on bluebox request):
```
.timer on
.eqp on
select count(1) from (select height from full_blocks where is_fully_compactified=0 order by random() limit 30);
select count(1) from (SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30);
```

The old stmt uses not optimal index (full_block_height) instead of is_fully_compactified.

Runtime on my main node (i5-8250U, db at local mSATA SSD):
```
sqlite> select count(1) from (SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30);
QUERY PLAN
|--CO-ROUTINE 1
|  |--SCAN TABLE full_blocks USING INDEX full_block_height
|  `--USE TEMP B-TREE FOR ORDER BY
`--SCAN SUBQUERY 1
30
Run Time: real 117.677 user 2.633940 sys 17.054738

sqlite> select count(1) from (select height from full_blocks where is_fully_compactified=0 order by random() limit 30);
QUERY PLAN
|--CO-ROUTINE 1
|  |--SEARCH TABLE full_blocks USING INDEX is_fully_compactified (is_fully_compactified=?)
|  `--USE TEMP B-TREE FOR ORDER BY
`--SCAN SUBQUERY 1
30
Run Time: real 0.286 user 0.112955 sys 0.007778
```

Some of the other mentioned runtimes are from members of my pool helping me to compare them with my own.
Because runtimes on my local db probably doesn't compare very well, as I changed the page_size to 64k for testing purpose.

**Review by arvidn — COMMENTED** [review]
Regarding the indices; we need to consider the migration of existing users. With this change, as I understand it, only new users would get the smaller index.

One way to migrate would be to give the new index a different name, leave stop creating the current index (for new users), and remove the current index in some future release.

**Comment by arvidn** [comment]
> Found and fixed the trailing whitespaces, but got a lots of other warnings and pre-commit didn't allow me to commit.
> 
> So I just changed it in web repo - I hope it passes all these checks now.

All the tests in pre-commit are run by the CI as well, so those have to be green.

**Comment by neurosis69** [comment]
> 
> Regarding the indices; we need to consider the migration of existing users. With this change, as I understand it, only new users would get the smaller index.
> 
> One way to migrate would be to give the new index a different name, leave stop creating the current index (for new users), and remove the current index in some future release.

1. index named "peak":

I really wouldn't bother much, because this index is already very small and I guess we are only talking about seconds which it would take on recreating the index. But I already added the comment to drop the old index, like you suggested in my PR #9112 

Rebuild duration on my side (i5-8250U, db at local mSATA SSD + 64K page size) was as follows:

```
sqlite> select * from sqlite_master where name='peak';
index|peak|block_records|19|CREATE INDEX peak on block_records(is_peak) where is_peak = 1
sqlite> select sum(pgsize) from dbstat where name='peak';
65536
sqlite> drop index peak;
sqlite> .timer on
sqlite> CREATE INDEX peak on block_records(is_peak);
Run Time: real 2.768 user 0.289760 sys 0.478361
sqlite> select sum(pgsize) from dbstat where name='peak';
10289152
Run Time: real 0.026 user 0.020756 sys 0.005531
sqlite> drop index peak;
Run Time: real 0.042 user 0.008816 sys 0.006275
sqlite> CREATE INDEX peak on block_records(is_peak) where is_peak = 1;
Run Time: real 0.234 user 0.071092 sys 0.128647
```

2. SELECT stmt (remove group by/having) 

The other change doesn't include any actual index change.
The query optimizer automatically takes the correct index when the stmt is changed.

3. Potential next index change

There is also a potential to change the index is_fully_compactified to a partial index like:
"CREATE INDEX IF NOT EXISTS is_fully_compactified on full_blocks(is_fully_compactified where is_fully_compactified=0"

But I haven't the time right now to test/verify if it would still be needed by other chia code.
So I'll probably make a followup PR with that.

File: chia/full_node/block_store.py [hunk 1]
```diff
@@ -58,7 +58,11 @@ async def create(cls, db_wrapper: DBWrapper):
         await self.db.execute("CREATE INDEX IF NOT EXISTS height on block_records(height)")
 
         await self.db.execute("CREATE INDEX IF NOT EXISTS hh on block_records(header_hash)")
-        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak)")
+
+        # this index was initially created and deployed without where condition "is_peak = 1"
+        # to replace the old index with the new one, uncomment the drop
+        # await self.db.execute("DROP INDEX IF EXISTS peak on block_records(is_peak)")
+        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak) where is_peak = 1")
```
Comment (arvidn) [inline chia/full_node/block_store.py:65]: I think the name of this index need to be something other than "peak", since that's the name of the current index. IF they have the same name, the commented-out `DROP INDEX IF EXISTS` would end up dropping both the new and the old index.

File: chia/wallet/wallet_block_store.py [hunk 2]
```diff
@@ -62,7 +62,12 @@ async def create(cls, db_wrapper: DBWrapper):
         await self.db.execute("CREATE INDEX IF NOT EXISTS height on block_records(height)")
 
         await self.db.execute("CREATE INDEX IF NOT EXISTS hh on block_records(header_hash)")
-        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak)")
+
+        # this index was initially created and deployed without where condition "is_peak = 1"
+        # to replace the old index with the new one, uncomment the drop
+        # await self.db.execute("DROP INDEX IF EXISTS peak on block_records(is_peak)")
+        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak) where is_peak = 1")
```
Comment (arvidn) [inline chia/wallet/wallet_block_store.py:69]: same comment here

**Comment by arvidn** [comment]
for reference, on my machine:

```
sqlite3 ~/.chia/mainnet/db/blockchain_v1_mainnet.sqlite
sqlite> .timer on
sqlite> drop index peak;
Run Time: real 24.339 user 0.102291 sys 0.296509
sqlite> CREATE INDEX peak on block_records(is_peak);
Run Time: real 984.614 user 7.466438 sys 34.449284
```

I had to drop the page cache to benchmark creation of the smaller index.

```
sqlite> drop index peak;
Run Time: real 10.683 user 0.036494 sys 0.042373
sqlite> CREATE INDEX peak on block_records(is_peak) where is_peak = 1;
Run Time: real 1008.655 user 6.667830 sys 31.874013
```

So, I don't think it's safe to assume creating the index is fast.

File: chia/full_node/block_store.py (hunk 1 above)
Comment (arvidn) [reply chia/full_node/block_store.py:65]: unless of course we don't have the ambition of upgrading existing users. given that it's potentially expensive to recreate the index, that argument could be made

**Comment by neurosis69** [comment]
> 
> 
> for reference, on my machine:
> 
> ```
> sqlite3 ~/.chia/mainnet/db/blockchain_v1_mainnet.sqlite
> sqlite> .timer on
> sqlite> drop index peak;
> Run Time: real 24.339 user 0.102291 sys 0.296509
> sqlite> CREATE INDEX peak on block_records(is_peak);
> Run Time: real 984.614 user 7.466438 sys 34.449284
> ```
> 
> I had to drop the page cache to benchmark creation of the smaller index.
> 
> ```
> sqlite> drop index peak;
> Run Time: real 10.683 user 0.036494 sys 0.042373
> sqlite> CREATE INDEX peak on block_records(is_peak) where is_peak = 1;
> Run Time: real 1008.655 user 6.667830 sys 31.874013
> ```
> 
> So, I don't think it's safe to assume creating the index is fast.

Very interesting. 

This would lead to some questions I have.
- which CPU your machine has and on which storage the db is located?
- was your db in sync since the beginning of mainnet? or was it synced later? (having fragmentation in mind)
- did you reorg it (vacuum) or is it untouched?
- whats the result of the following stmt on your machine?
`select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';`

On my machine
```
sqlite> select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';
97.0768199543746
Run Time: real 0.124 user 0.011323 sys 0.085690
```

I'll also ask other pool members to test, but I agree with what you said, even though I wouldn't say an index rebuild has to be done during a downtime. But it would lead to unnecessary questions.

tbh, I didn't expect that the timings you shared.

**Comment by arvidn** [comment]
my home directory is on an LVM volume combining 3 old spinning harddrives, circa 2006.
The machine is fast otherwise, as you can see from the timings, the vast majority of the time was spent in I/O wait.

**Comment by arvidn** [comment]
```
sqlite> select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';
67.8050152616412
```

**Comment by arvidn** [comment]
> was your db in sync since the beginning of mainnet? or was it synced later? (having fragmentation in mind)

Not quite from the start of mainnet, but almost.

> did you reorg it (vacuum) or is it untouched?

I have not done anything to it. i.e. only the full node has made changes to it (until just now when I took it offline to drop and create the index)

**Comment by neurosis69** [comment]
> 
> 
> ```
> sqlite> select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';
> 67.8050152616412
> ```

Can you verify if a vacuumed db on your setup would have significant better performance for index rebuilds?

In my opinion, the easiest way to achieve that, without destroying the state of your current db, would be with:
```
vacuum into "/home/chia/.chia/mainnet/db/blockchain_v1_mainnet.sqlite.backup"
.exit
sqlite3 /home/chia/.chia/mainnet/db/blockchain_v1_mainnet.sqlite.backup
select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';
.timer on
-- reexecute the drop/create index stmts for comparison
```

I'd expect significant better index rebuild times, even though the db is located at very old hdds and would compete for ios with your live db.

This topic we are disussing right now is only about, how to improve lookup performance and when to rebuild indexes for better client/full node performance, especially with focus on systems having db on storage like you. 
This may/should probably be part of different PR, because here I only wanted to propose the use of more efficient indexes/selects.

**Comment by neurosis69** [comment]
> 
> I had to drop the page cache to benchmark creation of the smaller index.
> 

I'm wondering. How did you drop the page caches?

**Comment by arvidn** [comment]
> I'm wondering. How did you drop the page caches?

as root:

```
echo 1 > /proc/sys/vm/drop_caches
```

**Comment by neurosis69** [comment]
> 
> 
> my benchmarks on making this index smaller (from a few weeks ago) suggested that there was no performance improvement, but I think saving space is also a worthy goal.

I've prepared a Pi4-8G with db located at ext4, 500GB USB2 HDD (manufactured in 2010) during the weekend.

Node was online, synced and syncing during these tests. 
The db was vacuumed with 4K page size (default which chia uses right now).

Regarding peak index:
```
sqlite> select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='block_records';
98.7786401932695
Run Time: real 28.645 user 0.533074 sys 4.048556
#clear fs caches
sqlite> CREATE INDEX peak on block_records(is_peak);
Run Time: real 5.950 user 1.002694 sys 1.614385
sqlite> drop index peak;
Run Time: real 0.573 user 0.049490 sys 0.057941
#clear fs caches
sqlite> CREATE INDEX peak on block_records(is_peak) where is_peak = 1;
Run Time: real 27.184 user 0.574478 sys 3.589601
#clear fs caches
```

Regarding query change:
```
sqlite> select sum(pgsize-unused)*100.0/sum(pgsize) from dbstat where name='full_blocks';
86.8481304255348
Run Time: real 451.642 user 3.547896 sys 52.424302
#clear fs caches
sqlite> select count(1) from (select height from full_blocks where is_fully_compactified=0 order by random() limit 30);
30
Run Time: real 4.925 user 0.471166 sys 0.166287
#clear fs caches
sqlite> select count(1) from (SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30);
30
Run Time: real 2714.864 user 14.129730 sys 146.152392
```
But I guess this query is only used on blueboxes? So it won't be executed on such a low-tier machine anyways?

On my drives I also found a old db backup from august, which I synced and I'm currently copying to the pi4 for a test.

**Comment by arvidn** [comment]
I think this change is good and should land. But the transition has to be considered. If the new index has the same name as the old one, there won't be a way (afaik) to know whether or not we have transitioned already. And I don't think it's feasible to drop and re-create the index on every startup.

**Comment by neurosis69** [comment]
got you. you don't have any naming conventions, do you?
I'd usually go with index names like 'is_peak_idx' or 'is_peak_idx_eq_1'

File: chia/full_node/block_store.py [hunk 3]
```diff
@@ -58,7 +58,11 @@ async def create(cls, db_wrapper: DBWrapper):
         await self.db.execute("CREATE INDEX IF NOT EXISTS height on block_records(height)")
 
         await self.db.execute("CREATE INDEX IF NOT EXISTS hh on block_records(header_hash)")
-        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak)")
+
+        # this index was initially created and deployed without where condition "is_peak = 1"
+        # to replace the old index with the new one, uncomment the drop
+        # await self.db.execute("DROP INDEX IF EXISTS peak on block_records(is_peak)")
```
Comment (arvidn) [inline chia/full_node/block_store.py:64]: I can't think of a reason to keep the old index. If we un-comment this line users will still be able to down-grade. There are no `INDEXED BY` clauses referring to this index by name, so downgrading will still just make sqlite use the new (smaller) index.

Unless you can think of a good reason to delay dropping this, I think we should do it right away in this patch.

File: chia/wallet/wallet_block_store.py [hunk 4]
```diff
@@ -62,7 +62,12 @@ async def create(cls, db_wrapper: DBWrapper):
         await self.db.execute("CREATE INDEX IF NOT EXISTS height on block_records(height)")
 
         await self.db.execute("CREATE INDEX IF NOT EXISTS hh on block_records(header_hash)")
-        await self.db.execute("CREATE INDEX IF NOT EXISTS peak on block_records(is_peak)")
+
+        # this index was initially created and deployed without where condition "is_peak = 1"
+        # to replace the old index with the new one, uncomment the drop
+        # await self.db.execute("DROP INDEX IF EXISTS peak on block_records(is_peak)")
```
Comment (arvidn) [inline chia/wallet/wallet_block_store.py:68]: same comment applies to this index. I think we could just drop it immediately

File: chia/full_node/block_store.py [hunk 5]
```diff
@@ -353,8 +357,7 @@ async def get_random_not_compactified(self, number: int) -> List[int]:
         # certain height are not compact. And if we do have compact orphan blocks, then all that
         # happens is that the occasional chain block stays uncompact - not ideal, but harmless.
         cursor = await self.db.execute(
-            f"SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 "
-            f"ORDER BY RANDOM() LIMIT {number}"
+            f"SELECT height FROM full_blocks WHERE is_fully_compactified=0 ORDER BY RANDOM() LIMIT {number}"
```
Comment (arvidn) [inline chia/full_node/block_store.py:360]: To be quite honest, I don't really understand this change. your new query looks simple and straight forward. I don't know why the existing code groups by `height` not why it's computing the sum of `is_fully_compactified`.

The original query was introduced [here](https://github.com/Chia-Network/chia-blockchain/commit/a59073fc38196134b66848c617d23df59b0db33a). To address [this](https://github.com/Chia-Network/chia-blockchain/pull/7175) issue.

perhaps @beechord has some comments

Comment (neurosis69) [reply chia/full_node/block_store.py:360]: The aim of this query, from what I understood is, that you just 
select a random `{number}` of heights for not yet compacitified rows (=`is_full_compactified=0`).

The main benefit of this change is, that the index on is_fully_compactified can be used instead of using index on height.

File: chia/full_node/block_store.py (hunk 3 above)
Comment (neurosis69) [reply chia/full_node/block_store.py:64]: That's what I also assumed, but then you brought [this ](https://github.com/Chia-Network/chia-blockchain/pull/9226#issuecomment-966985939) up.

A possible solution could be to drop the old index and create the new partial index in a non-blocking manner, like establish a separate function that autonomously executes on startup and is used for such index maintenance tasks. 
But I have no python experience, so this no function I could add to this PR myself.

Comment (arvidn) [reply chia/full_node/block_store.py:64]: Dropping the index might take ~25 seconds, and by not dropping it, we would save that startup time.

```
sqlite> drop index peak;
Run Time: real 24.339 user 0.102291 sys 0.296509
```

However, creating the new index may take as much as 1000 seconds.

```
sqlite> CREATE INDEX peak on block_records(is_peak) where is_peak = 1;
Run Time: real 1008.655 user 6.667830 sys 31.874013
```

So, creating the new index would dominate the cost in that case. Without dropping the old index, you also don't reap the benefits of the new index (afaict). In my measurements there's no speed improvement to use the smaller index. The benefit is a smaller database file. If you keep the old index, you don't get that improvement.

I suppose both dropping and creating the index could be done in a separate thread. The new one should probably be created first, before dropping the existing one. One risk with this is that spawning a thread and detaching it is risky, because we wouldn't be able to ensure it's done before shutting down. And just killing the thread while working on creating the index might be riskt (although, one would hope sqlite wouldn't create a half-baked index).

Comment (arvidn) [reply chia/full_node/block_store.py:64]: I'm not sure it's worth introducing that complexity just to improve the experience on a one-time startup delay though..

Comment (neurosis69) [reply chia/full_node/block_store.py:64]: > So, creating the new index would dominate the cost in that case. Without dropping the old index, you also don't reap the benefits of the new index (afaict). In my measurements there's no speed improvement to use the smaller index. The benefit is a smaller database file. If you keep the old index, you don't get that improvement.

In theory, there should be a minimal speed improvement due to less index maintenance: no need to maintain old peak in index, but yes, the main benefit I see is the reduced size on the long run.

Comment (neurosis69) [reply chia/full_node/block_store.py:64]: > I'm not sure it's worth introducing that complexity just to improve the experience on a one-time startup delay though..

Agree, this would only make sense if such an improvement could be reused on every startup and not only for this specific index but also for all others.

So I'll just drop the index immediately and create new one on startup for this PR.

File: chia/wallet/wallet_block_store.py (hunk 4 above)
Comment (neurosis69) [reply chia/wallet/wallet_block_store.py:68]: Regarding the how to drop indexes.
Did you check my suggested way to [drop indexes](https://github.com/Chia-Network/chia-blockchain/pull/9297/commits/520874105269f12ffbda365a7df5dc7d92d0010f) in my other PR #9297?

Instead of only using `DROP INDEX IF EXISTS peak`, I'd check if we are going to drop the index on the correct table and dynamically building the stmts.
Imho this would be a more robust and safe way to drop indexes.

File: chia/full_node/block_store.py (hunk 5 above)
Comment (beechord) [reply chia/full_node/block_store.py:360]: The basic problem here is, because consensus requires you to deal with forks and orphan blocks, blocks in the DB are not uniquely identified by height.

In my original PR, back when blueboxes still scanned work from the bottom up, I had to avoid naively compacting the oldest uncompact block because bluebox work would be wasted in multiple ways when compacting orphan blocks, first because `respond_compact_vdf` (iirc) would refuse to insert the compacted vdf for the orphan involved, second because even if you modified it, the network wouldn't gossip the proof because it belongs to an orphan that most people don't have. As this query is mostly run on bluebox timelords once every ~five minutes in the recommended settings, it's okay to spend a couple more seconds every five minutes if it means that you don't get stuck in an arbitrary part of the chain compacting proofs that nobody can use, at several minutes of CPU time per proof. And this was not just a theoretical problem but actually cropped up in practice by then. The more complicated query with the `HAVING` clause sidestepped all this.

A bit later in #8280, the query was modified again to return not the lowest uncompact block that was not definitely an orphan, but n random blocks that are not definitely orphans. I am not sure whether an index would even particularly help with the new query, as I believe `ORDER BY RANDOM()` computes a random number for every matching row to then choose those with the lowest, so again it is a scan through currently about 2/3 of the chain (https://www.arewecompactifiedyet.live/), index or not. Maybe a per-node random seed would help to make the ordering reproducible and therefore indexable, but that only really becomes necessary for a catchup protocol as detailed in #7678 where regular resource-constrained nodes run this query.
If you go and remove the `HAVING` here now, I believe but can't guarantee that blueboxing will basically keep working for now, but there will be more and more errors in the log as the network nears full compaction.

tl;dr this is not an easy fix, and it helps only a subset of nodes which is small and tends to consist of plotter-tier machines.

I've been busy with my dayjob recently so I don't know if there are even more current changes, but @fchirica knows the same things I do here and should be more up-to-speed than me.

Comment (neurosis69) [reply chia/full_node/block_store.py:360]: Very well explained. I didn't see the heights not beeing unique.

An improvement should still be possible by forcing the query to use index is_fully_compactified.
Example from my plotter:
```
sqlite> SELECT height FROM full_blocks indexed by is_fully_compactified GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30;
QUERY PLAN
|--SCAN TABLE full_blocks USING INDEX is_fully_compactified
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
Run Time: real 66.405 user 1.707902 sys 11.759515
...
sqlite> SELECT height FROM full_blocks GROUP BY height HAVING sum(is_fully_compactified)=0 order by random() limit 30;
QUERY PLAN
|--SCAN TABLE full_blocks USING INDEX full_block_height
`--USE TEMP B-TREE FOR ORDER BY
Run Time: real 140.793 user 2.701341 sys 18.591463
...
```

But as you mentioned, only high-tier nodes are effected.

Comment (beechord) [reply chia/full_node/block_store.py:360]: Hmm. Reading through the rest of the PR, it surprises me how fast `ORDER BY RANDOM()` is, so the currently unavoidable part is not as slow as I thought. If you can figure out a fast way of tagging orphan blocks for purposes of blueboxing, not consensus (so with relaxed requirements), you can index on is_orphan and height and hopefully get something reasonable.
The amount the right column is over 100k is the amount of orphans you have to deal with (the longer you stay online, the more you have):
```
sqlite> select (height - height%100000) as modheight,sum(is_fully_compactified),round(avg(length(block))),count() from full_blocks group by modheight;
0|100000|3425.0|100000
100000|99669|3426.0|100015
200000|99905|4474.0|101080
300000|49662|5056.0|100376
400000|13870|5219.0|100137
500000|12352|11862.0|100133
600000|12516|11793.0|100046
700000|12302|13666.0|100062
800000|9410|14518.0|100038
900000|8474|13404.0|100042
1000000|5640|17661.0|100247
1100000|1101|13196.0|63894
```

Comment (neurosis69) [reply chia/full_node/block_store.py:360]: I'm certainly missing some important information. 
How do you identify orphaned blocks using the current query?

I tried to reproduce on my side, but the only connection I see would be by matching prev_hash column from block_records table with header_hash column on full_blocks table.

Comment (beechord) [reply chia/full_node/block_store.py:360]: I don't - I merely know that when there's more than one block for a given height, all but one are orphans. I'm not aware of a way of identifying (as opposed to counting) orphans purely through a query, but I don't think I've ever tried to pull a table other than full_blocks into this. The way the rest of the implementation does it is for example by calling Blockchain.height_to_hash in FullNode._replace_proof: https://github.com/Chia-Network/chia-blockchain/blob/0d84578999c17c1ecd12dcc1bf4e0312a2f0b567/chia/full_node/full_node.py#L2078-L2082

If a consensus-strength is_orphan field existed, consensus code would need to make sure it was accurate and up-to-date every time a chain overtook another, which I presume is the reason it doesn't exist.

Comment (arvidn) [reply chia/full_node/block_store.py:360]: > If a consensus-strength is_orphan field existed, consensus code would need to make sure it was accurate and up-to-date every time a chain overtook another, which I presume is the reason it doesn't exist.

I believe it was a mistake to not include a field like this in the original schema. I haven't implemented one yet, but I intend to experiment with this (along with a bunch of other schema improvements) for a v2 database format. I was envisioning a field `in_main_chain` or something like that (i.e. the inverse of `is_orphaned`). Keeping it up to date should not be any more complex than the height-to-hash map we currently keep up to date.

Comment (neurosis69) [reply chia/full_node/block_store.py:360]: I have two suggestions how to proceed with the query in this PR.

1. Change it to just use the better index (is_fully_compactified)
```
SELECT height FROM full_blocks 
INDEXED BY is_fully_compactified 
GROUP BY height 
HAVING sum(is_fully_compactified)=0 
ORDER BY random() LIMIT 30;
```
2. As long as there is no column like `in_main_chain` available and if the main chain defines itself backwards by `prev_hash` from `max(height) from block_records`, then we could use a recursive CTE to get the uncompactified heights only of the main chain 
```
WITH RECURSIVE
	fb_blockchain(header_hash, height) AS
	(
		SELECT prev_hash, height-1 
                   FROM block_records 
                   WHERE height = (SELECT MAX(height) FROM block_records)
		UNION ALL
		SELECT block_records.prev_hash, block_records.height 
                   FROM block_records, fb_blockchain 
		   WHERE fb_blockchain.header_hash = block_records.header_hash 
		   ORDER BY block_records.height DESC
	)
SELECT fb.height 
   FROM fb_blockchain bc, full_blocks fb 
   WHERE bc.header_hash = fb.header_hash AND fb.is_fully_compactified=0 
   ORDER BY random() LIMIT 30;
```
Execution time on my plotter(with empty fs cache):
`Run Time: real 137.201 user 4.742898 sys 4.824835`

Execution time (without clearing fs cache):
`Run Time: real 5.522 user 3.670922 sys 1.828944`

Comment (arvidn) [reply chia/full_node/block_store.py:360]: A cheap way to check whether a block is in the chain or not is to use the height to hash map

Comment (neurosis69) [reply chia/full_node/block_store.py:360]: With my lack of python knowledge and time constraint, I don't see me figuring out how to use the height_to_hash map for this any time soon.

But I think the following SQL should do the work as well:
_(Fixed wrong height handling and used sligthly faster initial query)_
```
WITH RECURSIVE
	blockchain(header_hash, height) AS
	(
		SELECT prev_hash, height-1 
                   FROM block_records 
                   WHERE is_peak = 1
		UNION ALL
		SELECT block_records.prev_hash, blockchain.height-1 
                   FROM block_records, blockchain
		   WHERE blockchain.header_hash = block_records.header_hash
		   ORDER BY blockchain.height-1 DESC
	)
SELECT fb.height FROM blockchain bc, full_blocks fb 
	WHERE bc.header_hash = fb.header_hash AND fb.is_fully_compactified = 0 
	ORDER BY random() LIMIT 30;
```

**Comment by arvidn** [comment]
@neurosis69 after some internal discussion about this, we would like to make it configurable to opt-in to these data base "migrations", just to stay conservative. People that experience performance issues, will be able to enable these, without risking causing issues for others.

I intend to cherry-pick these changes and put them behind a configuration option. I hope you don't mind, I will be sure to attribute the changes to you!

File: chia/wallet/wallet_block_store.py (hunk 4 above)
Comment (arvidn) [reply chia/wallet/wallet_block_store.py:68]: it seems we're about to drop support for this wallet, in favor of the new. light-weight wallet.

**Comment by arvidn** [comment]
part1: https://github.com/Chia-Network/chia-blockchain/pull/9349

File: chia/full_node/block_store.py (hunk 5 above)
Comment (arvidn) [reply chia/full_node/block_store.py:360]: does this look right? https://github.com/Chia-Network/chia-blockchain/pull/9351
I attributed the commit to you

File: chia/wallet/wallet_block_store.py (hunk 4 above)
Comment (neurosis69) [reply chia/wallet/wallet_block_store.py:68]: nice, so I can play around with some new SQLs 🥳 
are the sources of the new light wallet already published?

**Comment by hoffmang9** [comment]
Light wallet is in this branch - https://github.com/Chia-Network/chia-blockchain/tree/protocol_and_cats_rebased

File: chia/full_node/block_store.py (hunk 5 above)
Comment (beechord) [reply chia/full_node/block_store.py]: Since I don't expect anybody to run a bluebox on a spinner, forcing index use looks fine to me. No objections from me. Eagerly awaiting an in_main_chain field, as I expect this to greatly simplify a compact-catchup implementation.

**Comment by neurosis69** [comment]
Obsoleted by light wallet
